tutorials/pubsub-spring-dedup-messages/index.md

```diff
@@ -0,0 +1,263 @@
+---
+title: Deduplicate Pub/Sub Messages Using Dataflow in a Spring Boot Application
+description: Use Dataflow's PubsubIO to deduplicate Pub/Sub messages in a Spring Boot application.
+author: anguillanneuf
+tags: PubSub, Spring, Spring Cloud GCP, Dataflow, Java
+date_published: 2020-02-29
+---
+
+
+Tianzi Cai | Developer Programs Engineer | Google Cloud
+
+
+## Objectives
+- Configure a Spring Boot application to use [Cloud Pub/Sub] as a message broker.
+- Use [Cloud Dataflow] to deduplicate messages.
+
+## Architecture
+![img](architecture.png)
+
+## Before You Begin
+
+1. Install the [Cloud SDK].
+
+1. Create a new Google Cloud project via the
+   [*New Project* page],
+   or via the `gcloud` command line tool.
+
+   ```shell script
+   export PROJECT_NAME=your-google-cloud-project-id
+   gcloud projects create $PROJECT_NAME
+   ```
+
+1. [Enable billing].
+
+1. Setup the Cloud SDK to your GCP project.
+
+   ```shell script
+   gcloud init
+   ```
+
+1. [Enable the APIs](https://console.cloud.google.com/flows/enableapi?apiid=dataflow,compute_component,pubsub): Dataflow, Compute Engine, Pub/Sub.
+
+1. Create a service account JSON key via the
+   [*Create service account key* page],
+   or via the `gcloud` command line tool.
+   Here is how to do it through the *Create service account key* page.
+
+   * From the **Service account** list, select **New service account**.
+   * In the **Service account name** field, enter a name.
+   * From the **Role** list, select **Project > Owner**.
+   * Click **Create**. Save this JSON file to a location on your computer.
+
+   Alternatively, you can use `gcloud` through the command line.
+
+   ```shell script
+   export PROJECT_NAME=$(gcloud config get-value project)
+   export SA_NAME=spring-app
+   export IAM_ACCOUNT=$SA_NAME@$PROJECT_NAME.iam.gserviceaccount.com
+
+   # Create the service account.
+   gcloud iam service-accounts create $SA_NAME --display-name $SA_NAME
+
+   # Set the role to Project Owner (*).
+   gcloud projects add-iam-policy-binding $PROJECT_NAME \
+     --member serviceAccount:$IAM_ACCOUNT \
+     --role roles/owner
+
+   # Create a JSON file with the service account credentials.
+   gcloud iam service-accounts keys create path/to/your/credentials.json \
+     --iam-account=$IAM_ACCOUNT
+   ```
+
+   > *Note:* The **Role** field authorizes your service account to access resources.
+   > You can view and change this field later by using the
+   > [GCP Console IAM page].
+   > If you are developing a production app, specify more granular permissions than **Project > Owner**.
+   > For more information, see
+   > [Granting roles to service accounts].
+
+   For more information, see
+   [Creating and managing service accounts].
+
+1. Set your `GOOGLE_APPLICATION_CREDENTIALS` environment variable to point to your service account key file.
```

Comment:
The following command directly sets the variable to the file, so we could probably shorten this to "Set the `GOOGLE_APPLICATION_CREDENTIALS` to your service account key file."